### ISSUE TITLE
Move reftest pass conditions into the manifest

### ISSUE THREAD

**Issue by gsnedders** [web-platform-tests/wpt#11285]
At the moment, we provide the reftest digraph in the manifest, but don't explicitly list the pass condition there, and as such every runner has to implement it itself. I'm not entirely convinced that wptrunner currently does this correctly, and further detaching it from test execution is probably a good idea.

I suggest, instead, that we provide an array of possible pass conditions, each pass condition being made up by an array of arrays of files that must render identically (i.e., split the set of URLs into a [quotient set](https://en.wikipedia.org/wiki/Equivalence_class)).

e.g.:

```
[
    [
        ["a.html", "b.html"]
    ]
]
```

This represents a single pass condition, where that pass condition is "a.html" and "b.html" must render identically.

```
[
    [
        ["a.html", "b.html"],
        ["c.html"]
    ]
]
```

This represents a single pass condition, where "a.html" and "b.html" must render identically, and "c.html" must render differently.

```
[
    [
        ["a.html", "b.html"],
        ["c.html"]
    ],
    [
        ["a.html", "b.html", "d.html"],
    ]
]
```

This represents two possible pass conditions: either a.html and b.html must render identically with c.html rendering differently *or* a.html, b.html, and d.html must render identically.

cc/ @jgraham @Hexcles @thejohnjansen

**Comment by jgraham** [comment]
I don't like the proposed format; the use of positional information to distinguish between match and mismatch as well as between different options makes it very hard to read. Also its not clear how one represents `a == b != c == d` because it only seems to allow one sequences of matches and one sequence of mismatches per option.

I see the rationale for moving this stuff into the manifest, but I'd like to see evidence that it doesn't significantly regress the time to create the manifest.

**Comment by gsnedders** [comment]
@jgraham Each pass condition is a disjoint set, partitioned into equivalent classes, so `a == b != c == d` becomes:

```
[
    ["a", "b"],
    ["c", "d"]
]
```

A more complex case would be `a == b != c != d`, which would become:

```
[
    ["a", "b"],
    ["c"],
    ["d"]
]

Though um this doesn't work, because `d == a` is valid there. Blargh.

The runner implementation would basically be:

```
function check_pass_condition(pass_condition) {
  let unequal = Object.create(null);
  for (let equiv of pass_condition) {
    let hash = get_screenshot(equiv[0]);
    if (hash in unequal)
      return false;
    unequal[hash] = true;
    for (let url of equiv.slice(1)) {
      if (hash != get_screenshot(url)) {
        return false;
      }
    }
  }
  return true;
}
```

**Comment by jgraham** [comment]
I should have read your comment more closely I guess :)

Anyway, yes, this still doesn't quite work. There is also the problem that it blows up rather for the case where you have deeply nested OR conditions, since you need to fully expand the path to each point in the tree.

**Comment by gsnedders** [comment]
My alternative suggestion now, and somewhat implemented:

An array of pass conditions, where each pass condition is an array of url_a, comp_type, url_b triples.

So, in a simply case with as a single OR:

```
      "pass_conditions": [
       [
        [
         "/css/CSS2/backgrounds/background-color-049.xht",
         "==",
         "/css/CSS2/backgrounds/background-color-049-020202-ref.xht"
        ]
       ],
       [
        [
         "/css/CSS2/backgrounds/background-color-049-030303-ref.xht",
         "==",
         "/css/CSS2/backgrounds/background-color-049.xht"
        ]
       ]
      ]
```

And the craziest case I know in the repo:

```
"pass_conditions": [
       [
        [
         "/css/css-transforms/transform3d-perspective-origin-ref.html",
         "==",
         "/css/css-transforms/transform3d-perspective-origin-001.html"
        ]
       ],
       [
        [
         "/css/css-transforms/transform3d-perspective-origin-001.html",
         "!=",
         "/css/css-transforms/transform-lime-square-ref.html"
        ]
       ],
       [
        [
         "/css/css-transforms/transform3d-perspective-origin-001.html",
         "!=",
         "/css/css-transforms/transform3d-rotatex-ref.html"
        ],
        [
         "/css/css-transforms/transform3d-rotatex-ref.html",
         "!=",
         "/css/css-transforms/transform-lime-square-ref.html"
        ]
       ],
       [
        [
         "/css/css-transforms/transform3d-rotatex-perspective-001.html",
         "!=",
         "/css/css-transforms/transform3d-rotatex-ref.html"
        ],
        [
         "/css/css-transforms/transform3d-perspective-origin-001.html",
         "!=",
         "/css/css-transforms/transform3d-rotatex-perspective-001.html"
        ],
        [
         "/css/css-transforms/transform3d-rotatex-ref.html",
         "!=",
         "/css/css-transforms/transform-lime-square-ref.html"
        ]
       ]
      ]
```

**Comment by Hexcles** [comment]
Sorry for my lack of context, but which part of the manifest are you talking about? I see `items.reftest`, which already has the passing conditions alongside with the references.

**Comment by gsnedders** [comment]
@Hexcles items.reftests currently contains the contents of the `<link rel=help>` elements from the file, which only forms the top-level ORs. Each of those references can then have links themselves. See https://web-platform-tests.org/writing-tests/reftests.html#complex-pass-conditions.

**Comment by foolip** [comment]
@gsnedders was this done with the manifest trie rewrite, or are pass conditions still not entirely contained in the manifest?

**Comment by gsnedders** [comment]
No, we never did this (although computing them is much simpler since the reftest simplification RFC).

**Comment by foolip** [comment]
I see, so does this mean that there's currently parsing of the tests as part of `./wpt run` to figure out the pass conditions?

**Comment by gsnedders** [comment]
> I see, so does this mean that there's currently parsing of the tests as part of `./wpt run` to figure out the pass conditions?

No, they're not parsed, the manifest contains (and always has contained) enough metadata to build up the pass conditions within it, but they're not explicitly given (which means any change to how we compute the set of pass conditions needs changes everywhere they're computed, which as far as I'm aware is currently in wptrunner, in Marionette, and presumably somewhere within Blink's web_tests).

**Comment by gsnedders** [comment]
I wonder if we should simply close this now; after the simplification it scarcely seems worthwhile?
